Q1: a majority of CpG dinucleotides are methylated.

Q2: Nanopore sequencing appears to provide better coverage. While coverage across the genome is relatively uniform, with both forms of sequencing sharing approximately 95.8% of reads, nanopore sequencing has a greater number of reads for more sites. 

Q3: Examining the violin plot, nanopore sequencing appears to have greater capability to detect changes in DNA methylation. When examining changes in methylation between normal and tumor cells, shifts in methylation patterns are much more apparent, particularly at highly methylated regions. In bisulfite sequencing, which may be less consistent in converting unmethylated cytosine to uracils, the proportion of reads that are 100% methylated are distinctly lower. 


Q4: Global tumorigenesis leads to an increase in DNA methylation. 


For Exercise 4:

Q5: In the region of interest highlighted in the PNG image (uploaded on GitHub), there is a significant increase in methylation observed in an apprixmately 3kb region from 25,251,000- 25,254,000 in the tumor cells. The increase in methylation may function to limit DNM3TA activity, influencing altered gene expression patterns that are characteristic of tumorigenesis. 

Q6: Gene imprinting occurs when gene expression is driven by inheritance from the parents. In imprinted genes, only one copy of the gene will be expressed, and the copy that is expressed is dictated by whether it's inherited from the mother or father. 

Q7: Phasing the reads results in splitting the sequencing into two haplotypes, identifying alleles that are either maternally or paternally inherited. 

Q8: No, not all reads can be phased. If there is no difference between the haplotypes, then no clustering can occur. Only genes in which there would be variation in the alleles inherited from each parent would have enough variation to enable clustering and differentiation between the two. 